Check product availability before validating orders

alegout opened the pull request:
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.3.x
| Description?  | BO: check products availabilities before validate order. Need to be merged after [BOOM-4361](https://github.com/PrestaShop/PrestaShop/pull/8607)
| Type?         | bug fix
| Category?     | BO
| BC breaks?    | no
| Deprecations? | Yes, Cart::isAllProductsInStock
| Fixed ticket? | [BOOM-4438](http://forge.prestashop.com/browse/BOOM-4438).
| How to test?  | Add a product with 5 in stock. Two customers add each 3 x the product. The two customers go to step 1, first one complete order, the second can not order because product is no more in stock.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/8752)
<!-- Reviewable:end -->

alegout commented:
Need to be merged after [BOOM-4361](https://github.com/PrestaShop/PrestaShop/pull/8607)

kpodemski commented on classes/PaymentModule.php:
```diff
@@ -212,6 +212,20 @@ public function validateOrder(
+            $productQuantity = Product::getQuantity($product['id_product'], $product['id_product_attribute']);
+
+            if ($productQuantity < 0 && !$availableOutOfStock) {
+                $psShoppingCart = Module::getInstanceByName('ps_shoppingcart');
```
this is so so so hardcoded... i don't think that's a good idea

Quetzacoalt91 replied:
Indeed, we can't accept so strong dependencies between the core and a module.

Did you want to redirect to the cart page?

alegout commented:
Review status: 0 of 10 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[classes/PaymentModule.php, line 222 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8752#-L5A-71lvVBFnR8qC3_4-r1-222:-L5LO2F2Z9uZ0enK0dq0:b-1eswgg) ([raw file](https://github.com/prestashop/prestashop/blob/9dc71e253df2d61fddd5714f9c7cddc72c2a1671/classes/PaymentModule.php#L222)):*
<details><summary><i>Previously, kpodemski (Krystian Podemski) wrote…</i></summary><blockquote>

this is so so so hardcoded... i don't think that's a good idea
</blockquote></details>

Hi Krystian, yes you are right. Before this, i thrown a PrestashopException, but there is no try/catch block to catch it... In result, user will get an error 500 without message. An other ticket is created to refacto this one, remove all die etc... (Thank you for your review)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8752)*
<!-- Sent from Reviewable.io -->

alegout commented:
Review status: 0 of 10 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[classes/PaymentModule.php, line 222 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8752#-L5A-71lvVBFnR8qC3_4-r1-222:-L5y09Cd6VrSomlJwMDM:b1uqm3s) ([raw file](https://github.com/prestashop/prestashop/blob/9dc71e253df2d61fddd5714f9c7cddc72c2a1671/classes/PaymentModule.php#L222)):*
<details><summary><i>Previously, alegout (Alban) wrote…</i></summary><blockquote>

Hi Krystian, yes you are right. Before this, i thrown a PrestashopException, but there is no try/catch block to catch it... In result, user will get an error 500 without message. An other ticket is created to refacto this one, remove all die etc... (Thank you for your review)
</blockquote></details>

http://forge.prestashop.com/browse/BOGOSS-88

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8752)*
<!-- Sent from Reviewable.io -->

LittleBigDev commented:
Reviewed 10 of 10 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[classes/Product.php, line 3595 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8752#-L7_1m07dSWnvOKjQtJ_:-L7_1m07dSWnvOKjQtJa:bmb3350) ([raw file](https://github.com/prestashop/prestashop/blob/9dc71e253df2d61fddd5714f9c7cddc72c2a1671/classes/Product.php#L3595)):*
> ```PHP
>     * @return int Available quantities
>     */
>     public static function getQuantity($id_product, $id_product_attribute = null, $cache_is_pack = null)
> ```

`$cache_is_pack` seems unused O_o

---

*[tests/Unit/Core/Cart/AbstractCartTest.php, line 182 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8752#-L7_AiEJ3MrtJ-TDLwEO:-L7_AiEJ3MrtJ-TDLwEP:bst1wau) ([raw file](https://github.com/prestashop/prestashop/blob/9dc71e253df2d61fddd5714f9c7cddc72c2a1671/tests/Unit/Core/Cart/AbstractCartTest.php#L182)):*
> ```PHP
> 
>         // Fix issue pack cache is set when adding products.
>         \Pack::resetStaticCache();
> ```

You should import `Pack` class to avoid using fully qualified names in the code

---

*[tests/Unit/Core/Cart/Adding/Product/AddPackTest.php, line 42 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8752#-L7_B0ji1qZ-AlmkWcO1:-L7_B0jj44d7VE7cWl4g:buwls3o) ([raw file](https://github.com/prestashop/prestashop/blob/9dc71e253df2d61fddd5714f9c7cddc72c2a1671/tests/Unit/Core/Cart/Adding/Product/AddPackTest.php#L42)):*
> ```PHP
>         $idProductInPack = $productPack->id;
>         $nbPack = \Product::getQuantity($idPack);
>         $nbProduct = \Product::getQuantity($idProductInPack);
> ```

You should import `Product` class to avoid using fully qualified names in the code

---

*[tests/Unit/Core/Cart/Adding/Product/AddPackTest.php, line 46 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8752#-L7_B9HwbG8tNNNn0cZz:-L7_B9HwbG8tNNNn0c_-:bst1wau) ([raw file](https://github.com/prestashop/prestashop/blob/9dc71e253df2d61fddd5714f9c7cddc72c2a1671/tests/Unit/Core/Cart/Adding/Product/AddPackTest.php#L46)):*
> ```PHP
>         $this->assertEquals(50, $nbProduct);
>         $this->assertTrue(\Pack::isInStock($idPack, 5));
>         $this->assertFalse(\Pack::isInStock($idPack, 6));
> ```

You should import `Pack` class to avoid using fully qualified names in the code

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8752)*
<!-- Sent from Reviewable.io -->

LittleBigDev reviewed (changes requested):
Please see Reviewable comments

alegout commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[classes/Product.php, line 3595 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8752#-L7_1m07dSWnvOKjQtJ_:-L7aY8rh78am3TFqxJLI:bfbm931) ([raw file](https://github.com/prestashop/prestashop/blob/9dc71e253df2d61fddd5714f9c7cddc72c2a1671/classes/Product.php#L3595)):*
<details><summary><i>Previously, LittleBigDev (Bastien Bieri) wrote…</i></summary><blockquote>

`$cache_is_pack` seems unused O_o
</blockquote></details>

yes, old parameter kept for compatibility

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8752)*
<!-- Sent from Reviewable.io -->

alegout commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[classes/PaymentModule.php, line 222 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8752#-L5A-71lvVBFnR8qC3_4-r1-222:-L88WxsbD5nm-HOEz_1x:b4lwtha) ([raw file](https://github.com/prestashop/prestashop/blob/9dc71e253df2d61fddd5714f9c7cddc72c2a1671/classes/PaymentModule.php#L222)):*
<details><summary><i>Previously, Quetzacoalt91 (Thomas N) wrote…</i></summary><blockquote>

Indeed, we can't accept so strong dependencies between the core and a module.

Did you want to redirect to the cart page?
</blockquote></details>

Yes indeed, do you have a better way? thank you :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8752)*
<!-- Sent from Reviewable.io -->

alegout committed:
Check products availability before ordering

Check products availability before ordering

Fix issue with url rewriting

alegout commented:
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[tests/Unit/Core/Cart/AbstractCartTest.php, line 182 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8752#-L7_AiEJ3MrtJ-TDLwEO:-L8Xv3LbApqPVHPqiVbI:b-896fix) ([raw file](https://github.com/prestashop/prestashop/blob/9dc71e253df2d61fddd5714f9c7cddc72c2a1671/tests/Unit/Core/Cart/AbstractCartTest.php#L182)):*
<details><summary><i>Previously, LittleBigDev (Bastien Bieri) wrote…</i></summary><blockquote>

You should import `Pack` class to avoid using fully qualified names in the code
</blockquote></details>

Done.

---

*[tests/Unit/Core/Cart/Adding/Product/AddPackTest.php, line 42 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8752#-L7_B0ji1qZ-AlmkWcO1:-L8Xv2N_2gW6C2newPkf:b-896fix) ([raw file](https://github.com/prestashop/prestashop/blob/9dc71e253df2d61fddd5714f9c7cddc72c2a1671/tests/Unit/Core/Cart/Adding/Product/AddPackTest.php#L42)):*
<details><summary><i>Previously, LittleBigDev (Bastien Bieri) wrote…</i></summary><blockquote>

You should import `Product` class to avoid using fully qualified names in the code
</blockquote></details>

Done.

---

*[tests/Unit/Core/Cart/Adding/Product/AddPackTest.php, line 46 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8752#-L7_B9HwbG8tNNNn0cZz:-L8Xv2jM36KUK_NOYSA5:b-896fix) ([raw file](https://github.com/prestashop/prestashop/blob/9dc71e253df2d61fddd5714f9c7cddc72c2a1671/tests/Unit/Core/Cart/Adding/Product/AddPackTest.php#L46)):*
<details><summary><i>Previously, LittleBigDev (Bastien Bieri) wrote…</i></summary><blockquote>

You should import `Pack` class to avoid using fully qualified names in the code
</blockquote></details>

Done.

---

*[classes/PaymentModule.php, line 222 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8752#-L5A-71lvVBFnR8qC3_4-r1-222:-L8aqaWj_vNn3twrmEE9:b-896fix) ([raw file](https://github.com/prestashop/prestashop/blob/9dc71e253df2d61fddd5714f9c7cddc72c2a1671/classes/PaymentModule.php#L222)):*
<details><summary><i>Previously, alegout (Alban) wrote…</i></summary><blockquote>

Yes indeed, do you have a better way? thank you :)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8752)*
<!-- Sent from Reviewable.io -->

alegout commented:
changes done, thank you @LittleBigDev @kpodemski @Quetzacoalt91 @rGaillard

LittleBigDev commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[tests/Unit/Core/Cart/Adding/Product/AddPackTest.php, line 201 at r3](https://beta.reviewable.io/reviews/prestashop/prestashop/8752#-L8bu_UDA-wpQLbpPbIi:-L8bu_UDA-wpQLbpPbIj:b-fhie9y) ([raw file](https://github.com/prestashop/prestashop/blob/ed4cfdc1444aae65234522922ccb13fcdee2981e/tests/Unit/Core/Cart/Adding/Product/AddPackTest.php#L201)):*
> ```PHP
>         $this->assertTrue($this->cart->updateQty(51, $this->productInPack->id));
>         $outOfStock = StockAvailable::outOfStock($this->productInPack->id);
>         $this->assertEquals(1, $this->pack->isAvailableWhenOutOfStock($outOfStock));
> ```

Think about adding it in dev doc because the whole test is hard to understand

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/prestashop/prestashop/8752#-:-L8bv3cF-XYDp_wZgpYR:bnfp4nl)*
<!-- Sent from Reviewable.io -->

mickaelandrieu commented:
Hello @alegout,

this contribution doesn't pass travis tests, would you mind to rebase and take a look?

alegout commented:
It's fixed, thank you @mickaelandrieu

marionf commented:
@alegout 

When I click on "Continue" after the carrier selection, if the product is out of stock I am redirected to the home page. I should be redirected to the cart with the error message below:

![capture du 2018-04-05 10-54-39](https://user-images.githubusercontent.com/13449658/38356341-c47fc622-38bf-11e8-9e2c-5169e201039a.png)

I still can order an out of stock product because there isn't a check on the last step when I click on "order with an obligation to pay", if we can find a way to do it, it will be awesome :)

marionf commented:
@alegout 

If I allow order for out of stock product, I am redirected to the cart and can't complete the order

alegout committed:
Code refacto, fix issue ordering product out of stock

Code refacto, fix issue ordering product out of stock

Fix issue with product quantity

marionf commented:
@alegout 
I have a product with combinations (300 in stock). The default behavior is to allow order of out of stock product. Except for my product with combinations, I choose to refuse the order of out of stock product.
With the first account I order 200 products, with the second account I order also 200 products of the same combination. It works while it shouldn't, I have -100 in stock.

![capture du 2018-04-06 17-06-27](https://user-images.githubusercontent.com/13449658/38428788-e9acf35e-39bc-11e8-9513-85f5d3dc8b39.png)

alegout commented:
![image](https://media.giphy.com/media/hppWdK8gcmzXq/giphy.gif)

alegout committed:
Fix issue with products quantities

Fix issue with products quantities

marionf commented:
@alegout, little issue with a customizable product 
![capture du 2018-04-09 16-40-28](https://user-images.githubusercontent.com/13449658/38504984-6c221764-3c16-11e8-95b8-19715628cc94.png)

alegout committed:
Correctly get quantity of a customizable product. Code refacto.

alegout committed:
Fix issue with product quantity query

eternoendless commented:
Unit tests are failing

eternoendless commented:
Reviewed 3 of 5 files at r3, 1 of 1 files at r4, 3 of 3 files at r7.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[classes/checkout/CheckoutPaymentStep.php, line 50 at r7](https://beta.reviewable.io/reviews/prestashop/prestashop/8752#-L9z1vEi2lDThBcw5lA9:-L9z1vEi2lDThBcw5lAA:b-kbqh9k) ([raw file](https://github.com/prestashop/prestashop/blob/b46edde375484467f010a2563971c7f15e6d42d4/classes/checkout/CheckoutPaymentStep.php#L50)):*
> ```PHP
>         $allProductsInStock = $this->getCheckoutSession()->getCart()->isAllProductsInStock();
> 
>         if ($allProductsInStock != true) {
> ```

Please either use strict comparison `!== true` or simply `!$allProductsInStock`

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/prestashop/prestashop/8752)*
<!-- Sent from Reviewable.io -->

tomlev commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[classes/Cart.php, line 4705 at r7](https://beta.reviewable.io/reviews/prestashop/prestashop/8752#-LADvwXI9J1PF3_UCdq-:-LADvwXJ0d3VKFOl8eLw:b6stgt) ([raw file](https://github.com/prestashop/prestashop/blob/b46edde375484467f010a2563971c7f15e6d42d4/classes/Cart.php#L4705)):*
> ```PHP
>             );
> 
>             if (!$exclusive
> ```

an error is triggered below if $exclusive is given, so you can always consider $exclusive value is false (default)

---

*[classes/Cart.php, line 4716 at r7](https://beta.reviewable.io/reviews/prestashop/prestashop/8752#-LADwABO77nv1CbjSnEE:-LADwABO77nv1CbjSnEF:b-ym7ogq) ([raw file](https://github.com/prestashop/prestashop/blob/b46edde375484467f010a2563971c7f15e6d42d4/classes/Cart.php#L4716)):*
> ```PHP
>                 }
> 
>                 if ($productInStock > 0 && $productOutOfStock > 0) {
> ```

with default exclusive value this can be simplified ;)

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/prestashop/prestashop/8752)*
<!-- Sent from Reviewable.io -->

tomlev commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[classes/Cart.php, line 4705 at r7](https://beta.reviewable.io/reviews/prestashop/prestashop/8752#-LADvwXI9J1PF3_UCdq-:-LAHVqNg8KHHk5viFIZE:bxc0eo5) ([raw file](https://github.com/prestashop/prestashop/blob/b46edde375484467f010a2563971c7f15e6d42d4/classes/Cart.php#L4705)):*
<details><summary><i>Previously, tomlev (Thomas LEVIANDIER) wrote…</i></summary><blockquote>

an error is triggered below if $exclusive is given, so you can always consider $exclusive value is false (default)
</blockquote></details>

my bad  if not catched error is not blocking code execution

---

*[classes/Cart.php, line 4716 at r7](https://beta.reviewable.io/reviews/prestashop/prestashop/8752#-LADwABO77nv1CbjSnEE:-LAHW2Gu2H90PmcfxKbs:b32wtd) ([raw file](https://github.com/prestashop/prestashop/blob/b46edde375484467f010a2563971c7f15e6d42d4/classes/Cart.php#L4716)):*
<details><summary><i>Previously, tomlev (Thomas LEVIANDIER) wrote…</i></summary><blockquote>

with default exclusive value this can be simplified ;)
</blockquote></details>

ok

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/prestashop/prestashop/8752)*
<!-- Sent from Reviewable.io -->

tomlev commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[classes/Cart.php, line 4705 at r7](https://beta.reviewable.io/reviews/prestashop/prestashop/8752#-LADvwXI9J1PF3_UCdq-:-LAHW7rh2Hz3kp-G6QDd:b32wtd) ([raw file](https://github.com/prestashop/prestashop/blob/b46edde375484467f010a2563971c7f15e6d42d4/classes/Cart.php#L4705)):*
<details><summary><i>Previously, tomlev (Thomas LEVIANDIER) wrote…</i></summary><blockquote>

my bad  if not catched error is not blocking code execution
</blockquote></details>

ok

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/prestashop/prestashop/8752)*
<!-- Sent from Reviewable.io -->

alegout committed:
Fix issue with product quantity

marionf commented:
@alegout 

I have a standard product "A" with a stock of 60
I have a pack "B" of 10 products A, with a stock of 6
I choose to decrement the pack only.
In 2 browsers I add 30 products A and 3 packs B to the cart.
I can validate the 2 orders (this is ok) but I have a fatal error after the validation:
![capture du 2018-04-18 12-07-04](https://user-images.githubusercontent.com/13449658/38932706-4a7e68e4-4317-11e8-8251-f9a51f531e7a.png)
Order is well created.

Now I choose to decrement products in pack only.
In 2 browsers I add 3 packs B and 1 product A.
I can validate the first order (it's ok)
I can't validate the second order, I am well redirected to the cart but the error message isn't displayed

marionf commented:
The first issue is solved by this [PR](https://github.com/PrestaShop/PrestaShop/pull/8960)

eternoendless commented:
Rebased to include the fix from https://github.com/PrestaShop/PrestaShop/pull/8960

tomlev committed:
fix wrong quantity message display

eternoendless commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r9, 1 of 1 files at r10, 2 of 2 files at r11.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/8752#-:-LAwBom2AcM5EF84-NLs:bnfp4nl)*
<!-- Sent from Reviewable.io -->

marionf commented:
Still the same issue with the error in cart not displayed

tomlev committed:
fix available pack quantity calculation

tomlev committed:
fix quantity detection on cart page load

LittleBigDev commented:
Reviewed 2 of 2 files at r12.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[classes/Cart.php, line 4713 at r12](https://reviewable.io/reviews/prestashop/prestashop/8752#-LB-r5Ca5n_ndr9fvF1O:-LB-r5Ca5n_ndr9fvF1P:b-15pxm5) ([raw file](https://github.com/prestashop/prestashop/blob/02117bdfce866dc35dbfc259cbc017af59b689a5/classes/Cart.php#L4713)):*
> ```PHP
> 
>             if (!$exclusive
>                 && ($productQuantity < 0 && !$availableOutOfStock)
> ```

weird formatting here

---


*Comments from [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/8752)*
<!-- Sent from Reviewable.io -->

LittleBigDev commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[classes/Cart.php, line 4713 at r12](https://reviewable.io/reviews/prestashop/prestashop/8752#-LB-r5Ca5n_ndr9fvF1O:-LB06XKS81bl8oLd7KLy:bjw398f) ([raw file](https://github.com/prestashop/prestashop/blob/02117bdfce866dc35dbfc259cbc017af59b689a5/classes/Cart.php#L4713)):*
<details><summary><i>Previously, LittleBigDev (Bastien Bieri) wrote…</i></summary><blockquote>

weird formatting here
</blockquote></details>

Nevermind, I didn't notice the parenthesis

---


*Comments from [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/8752#-:-LB06kWC56Qqjj2Us1jS:bnfp4nl)*
<!-- Sent from Reviewable.io -->

LittleBigDev reviewed (approved):
Thanks @tomlev

eternoendless commented:
Reviewed 2 of 2 files at r12.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/8752)*
<!-- Sent from Reviewable.io -->

eternoendless commented:
Thank you @alegout and @tomlev !

sirius82 commented:
Hi everybody ! I'm sorry, I don't understand very well what I have to do. PS 1.7.3.0 in general settings, orders are not allowed when stock = 0. But it's possible to buy and pay a product without stock. (stock = -1). 
Thanks for your help.

reanim8ed commented:
@sirius82 upgrade to 1.7.3.2 it is fixed in that release

jefkeijers commented:
Does this fix also affect when an order is created through the API? Currently one of our clients sends their orders in csv-format, we prepare the orders and afterwards we import them in our Prestashop. It may occur that a product will get a negative stock but this is no problem as the order is already picked.
